### DISCUSSION TITLE
How to derive a public key from private key in ed25519 scheme using openssl 3.0?

### DISCUSSION THREAD

**Discussion by ToolmanP** [openssl/openssl#21865] [Q&A] [answered]
To begin with, I've searched a lot of documents the openssl provided and the issues down the repository. None of them seems to provide a concise solution to this.

I know i can generate private key and sign the message by the following snippet,

```c++
  EVP_PKEY *priv = nullptr;
  auto pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, NULL);

  if (pctx == NULL) {
    ERR_print_errors_fp(stdout);
    return;
  }

  if (EVP_PKEY_keygen_init(pctx) <= 0) {
    ERR_print_errors_fp(stdout);
    return;
  }

  if (EVP_PKEY_keygen(pctx, &priv) <= 0) {
    ERR_print_errors_fp(stdout);
    return;
  }

  auto md_ctx = EVP_MD_CTX_new();

  if (int status = EVP_DigestSignInit_ex(md_ctx, nullptr, nullptr, nullptr,
                                         nullptr, priv, nullptr);
      status <= 0) {
    ERR_print_errors_fp(stdout);
  }

  size_t sig_len;
  if (iEVP_DigestSign(md_ctx, nullptr, &sig_len,
                                  reinterpret_cast<const uint8_t *>(msg.data()),
                                  msg.length()) <= 0) {
    ERR_print_errors_fp(stdout);
  }
  std::string sig(sig_len, 0);

  if (EVP_DigestSign(
          md_ctx, reinterpret_cast<uint8_t *>(sig.data()), &sig_len,
          reinterpret_cast<const uint8_t *>(msg.data()), msg.length()) <=  0) {
    ERR_print_errors_fp(stdout);
  }


```

I know i can derive a ec-point representing the public key using the generic $pub = priv * G$ method and i was able to get the struct BIGNUM by extracting the octet string from the `priv` and construct one. But the question is i cannot get a ec-point group from NID_ED25519. So How can I possibly verify the signature without that priv key since we only need the public key to verify the signature?
I'll thoroughly appreciate it if someone can help it.

**Comment by nhorman** [comment]
I believe you're asking 2 questions here:

1) How can I extract an EC_GROUP value given in EC_POINT input?
2) How can I validate a signature using an EC_POINT (public key) input

The answer to (1), prior to openssl 3.0 was to use EVP_PKEY_get_EC_KEY, followed by EC_KEY_get0_group, but those methods I believe are deprecated.  The openssl 3.0 method I believe would be to use EC_group_new_from_params, extracting the needed parameters from the keypair you have.

The answer to (2), is that you shouldn't need an EC_GROUP value when validating a signature.  Ostensibly, when you have a message you need to verify, you'll receive the senders public key in PEM or DER format, from which you will build a key using PEM_read_bio_PUBKEY, which will return an EVP_PKEY to you, containing all the public key material needed for validation.  That will be fed as an input to EV_PKEY_DigestVerify_init, which will then be used in the validation

Hope that helps

**Reply by mattcaswell** [reply]
> The answer to (1), prior to openssl 3.0 was to use EVP_PKEY_get_EC_KEY, followed by EC_KEY_get0_group, but those methods I believe are deprecated. The openssl 3.0 method I believe would be to use EC_group_new_from_params, extracting the needed parameters from the keypair you have.

This doesn't work with ED25519. The type of curve being used is different and is not represented via EC_GROUP, and the key is not stored in an EC_KEY. It uses a completely separate implementation.

**Comment by t8m** [comment, marked as answer]
For ED25519, the public key is generated along with the private key. You should be able to use the same EVP_PKEY to verify the signature.

If you want to extract the public key only you can for example use EVP_PKEY_todata() with the EVP_PKEY_PUBLIC_KEY to output the public key data as OSSL_PARAMs and then create a new (public key only) EVP_PKEY with EVP_PKEY_fromdata() from these params.

**Reply by slontis** [reply]
EVP_PKEY_new_raw_public_key_ex() can also be used
